### PR TITLE
Feature/8504 Open test certificate from test result

### PR DESCRIPTION
**Pull request by naveeddotio**
## Description
Functionality to open test certificate directly from test result screen

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8504

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 1]
```diff
@@ -668,6 +687,173 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
+							}
+							self.healthCertificateService.removeHealthCertificate(healthCertificate)
+							let isPersonStillExistent = self.healthCertificateService.healthCertifiedPersons.value.contains(healthCertifiedPerson)
+
+							// Only pop to root if we did not removed the last certificate of a person (because this removes the person, too). A pop would trigger a reload of content which was removed before. If so, dismiss to go back to certificate overview.
+							if isPersonStillExistent {
+								self.navigationController?.popToRootViewController(animated: true)
+							} else {
+								self.dismiss()
+							}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:766]: I think it makes sense to always pop back to the test result on deletion. And probably not back to root, as there could be other view controllers on the stack still.

```suggestion
							self.navigationController?.popViewController(animated: true)
```

File: src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift [hunk 2]
```diff
@@ -468,6 +470,31 @@ class CoronaTestService {
...
+		var healthTuple: HealthCertificateTuple?
+		self.healthCertificateService.healthCertifiedPersons.value.forEach { healthCertifiedPerson in
+			healthCertifiedPerson.healthCertificates.forEach { healthCertificate in
+				if healthCertificate.uniqueCertificateIdentifier == uniqueCertificateIdentifier {
+					healthTuple = HealthCertificateTuple(certificate: healthCertificate, certifiedPerson: healthCertifiedPerson)
+				}
+			}
+		}
+		return healthTuple
+	}
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift:496]: I think if you just need it here you can just actually use a tuple. And we don't need the `get` prefix in Swift 😉

```suggestion
	func healthCertificateTuple(for uniqueCertificateIdentifier: String) -> (certificate: HealthCertificate, certifiedPerson: HealthCertifiedPerson)? {
		var healthTuple: (certificate: HealthCertificate, certifiedPerson: HealthCertifiedPerson)?
		self.healthCertificateService.healthCertifiedPersons.value.forEach { healthCertifiedPerson in
			healthCertifiedPerson.healthCertificates.forEach { healthCertificate in
				if healthCertificate.uniqueCertificateIdentifier == uniqueCertificateIdentifier {
					healthTuple = (certificate: healthCertificate, certifiedPerson: healthCertifiedPerson)
				}
			}
		}
		return healthTuple
	}
```

**Review by nickguendling — CHANGES_REQUESTED** [review]

**Review by pascal-brause — CHANGES_REQUESTED** [review]
I see an dependency to the pdf export story, because the flows would differ now.
See https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-8384 for more details. 
Perhaps we have to use the changed UI in the certificate details also here.

**Review by pascal-brause — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/HealthCertificate/HealthCertificateViewController.swift [hunk 3]
```diff
@@ -44,6 +44,10 @@ class HealthCertificateViewController: UIViewController, UITableViewDataSource,
 		setupViewModel()
 	}
 
+	override func viewWillDisappear(_ animated: Bool) {
+		
+	}
+
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/HealthCertificate/HealthCertificateViewController.swift:50]: Whoops, whats that? 😆

File: src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift [hunk 4]
```diff
@@ -468,6 +470,31 @@ class CoronaTestService {
...
+			if self.pcrTest?.registrationToken == testCertificateRequest.registrationToken {
+				pcrTest?.uniqueCertificateIdentifier = uniqueCertificateIdentifier
+			}
+		case .antigen:
+			if self.antigenTest?.registrationToken == testCertificateRequest.registrationToken {
+				self.antigenTest?.uniqueCertificateIdentifier = uniqueCertificateIdentifier
+			}
+		}
+	}
+
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift:485]: Should be moved to the private section

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 5]
```diff
@@ -16,7 +16,8 @@ class ExposureSubmissionTestResultViewModel {
 		onContinueWithSymptomsFlowButtonTap: @escaping () -> Void,
 		onContinueWarnOthersButtonTap: @escaping (@escaping (Bool) -> Void) -> Void,
 		onChangeToPositiveTestResult: @escaping () -> Void,
-		onTestDeleted: @escaping () -> Void
+		onTestDeleted: @escaping () -> Void,
+		onTestCertificateTapped: @escaping(HealthCertificate, HealthCertifiedPerson) -> Void
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:20]: Just realized, would be more in line with the others if named like this

```suggestion
		onTestCertificateCellTap: @escaping(HealthCertificate, HealthCertifiedPerson) -> Void
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 6]
```diff
@@ -25,6 +26,7 @@ class ExposureSubmissionTestResultViewModel {
 		self.onContinueWarnOthersButtonTap = onContinueWarnOthersButtonTap
 		self.onChangeToPositiveTestResult = onChangeToPositiveTestResult
 		self.onTestDeleted = onTestDeleted
+		self.onTestCertificateTapped = onTestCertificateTapped
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:29]:
```suggestion
		self.onTestCertificateCellTap = onTestCertificateCellTap
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 7]
```diff
@@ -120,6 +122,7 @@ class ExposureSubmissionTestResultViewModel {
 
 	private let onChangeToPositiveTestResult: () -> Void
 	private let onTestDeleted: () -> Void
+	private let onTestCertificateTapped: (HealthCertificate, HealthCertifiedPerson) -> Void
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:125]:
```suggestion
	private let onTestCertificateCellTap: (HealthCertificate, HealthCertifiedPerson) -> Void
```

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift [hunk 8]
```diff
@@ -539,7 +542,31 @@ extension ExposureSubmissionTestResultViewModel {
...
 		
-		var cells = [DynamicCell.title2(
+		var cells = [DynamicCell]()
+
+		if coronaTest.certificateRequested, let healthTuple = coronaTestService.healthCertificateTuple(for: coronaTest.uniqueCertificateIdentifier ?? "") {
+			
+			cells.append(DynamicCell.identifier(
+				ExposureSubmissionTestResultViewController.CustomCellReuseIdentifiers.healthCertificateCell,
+				action: .execute { _, _ in
+					self.onTestCertificateTapped(healthTuple.certificate, healthTuple.certifiedPerson)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/ExposureSubmissionTestResultViewModel.swift:552]:
```suggestion
					self.onTestCertificateCellTap(healthTuple.certificate, healthTuple.certifiedPerson)
```

**Review by nickguendling — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/HealthCertificate/HealthCertificateViewController.swift (hunk 3 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/HealthCertificates/HealthCertificate/HealthCertificateViewController.swift:50]:
```suggestion
```

**Pull request merged**